gufe/protocols/protocol_settings.py

```diff
@@ -0,0 +1,381 @@
+
+    TODO
+    ----
+    * In the future make precision and deterministic forces user defined too.
+    """
+    precision: Optional[float] = None
+
+class OpenMMEngineSettings(BaseModel):
+    """OpenMM MD engine settings
+
+    Attributes
+    ----------
+    compute_platform : str, optional
+      Which compute platform to perform the simulation on. If None, the
+      fastest compute platform available will be chosen. Default None.
+
+    TODO
+    ----
+    * In the future make precision and deterministic forces user defined too.
+    """
+    compute_platform: Optional[str] = None
+
+
+class SamplerSettings(BaseModel):
+    """Settings for the Equilibrium sampler, currently supporting either
+    HybridSAMSSampler or HybridRepexSampler.
+
+    Attributes
+    ----------
+    sampler_method : str
+      Sampler method to use, currently supports repex (hamiltonian replica
+      exchange) and sams (self-adjusted mixture sampling). Default repex.
+    online_analysis_interval : int
+      The interval at which to perform online analysis of the free energy.
+      At each interval the free energy is estimate and the simulation is
+      considered complete if the free energy estimate is below
+      ``online_analysis_target_error``. Default `None`.
+    online_analysis_target_error : float * unit.boltzmann_constant * unit.kelvin
+      Target error for the online analysis measured in kT.
+      Once the free energy is at or below this value, the simulation will be
+      considered complete.
+    online_analysis_minimum_iterations : float
+      Set number of iterations which must pass before online analysis is
+      carried out. Default 50.
+    flatness_criteria : str
+      SAMS only. Method for assessing when to switch to asymptomatically
+      optimal scheme.
+      One of ['logZ-flatness', 'minimum-visits', 'histogram-flatness'].
+      Default 'logZ-flatness'.
+    gamma0 : float
+      SAMS only. Initial weight adaptation rate. Default 0.0.
+
+    TODO
+    ----
+    * Work out how this fits within the context of independent window FEPs.
+    * It'd be great if we could pass in the sampler object rather than using
+      strings to define which one we want.
+    """
+    class Config:
+        arbitrary_types_allowed = True
+
+    sampler_method = "repex"
+    online_analysis_interval: Optional[int] = None
+    online_analysis_target_error = 0.2 * unit.boltzmann_constant * unit.kelvin
+    online_analysis_minimum_iterations = 50
+    flatness_criteria = 'logZ-flatness'
+    gamma0 = 0.0
+
+    @validator('online_analysis_target_error',
+               'online_analysis_minimum_iterations', 'gamma0')
+    def must_be_positive(cls, v):
+        if v < 0:
+            errmsg = ("Online analysis target error, minimum iteration "
+                      "and SAMS gamm0 must be 0 or positive values")
+            raise ValueError(errmsg)
+        return v
+
+
+class BarostatSettings(BaseModel):
```

Comment:
I think this is a problematic design, since it convolves parameters that define the alchemical state and hence impact the free energy (`pressure`) with parameters that simply control the efficiency of sampling (`frequency`).

I highly recommend instead following something like what `openmmtools` does and defining a [`ThermodynamicState`](https://github.com/choderalab/openmmtools/blob/main/openmmtools/states.py#L385) that encodes the information about thermodynamic parameters that will impact the thermodynamic ensemble sampled (and hence the free energy)---such as `temperature` and optionally `pressure`, `pH`, etc. 

This should be separated from settings that simply impact the efficiency of the sampling.

Reply:
Starting with a point of curiosity - I realise that at the limit of infinite sampling the MC frequency has no impact on how the thermodynamic state of the system, but given we are likely running short simulations and often non-equilibrium ones, does the change in relaxation time not affect the sampled free energy?

A thought on the splitting of parameters - I can understand & support the idea of separating out the "things that affect the ensemble", but at what limit can we differentiate between algorithmic details that can have an impact and broad ensemble parameters? (e.g. the choice of `gamma_ln` for Langevin integrators or the splitting sequence, etc...)

Reply:
There will be many things that impact the the finite-time behavior of simulations---up to and including the exact choice of random numbers you generate. But it simply isn't sensible to convolve (1) parameters that impact the infinite-time limit of the calculation (thermodynamic state, potential function) with (2) parameters that impact the rate at which bias and variance (and hence total error to the asumptotic limit) diminish. There is a very good reason for this: 

First, it's possible to independently optimize and tune (or even auto-tune adaptively on the fly) the simulation strategy to make the rate of convergence to the asymptotic limit, and hence make the calculation more efficient. That makes for a very useful separation of effort, since work can be carried out independently on optimizing the efficiency of protocols, alchemical methods, engines, etc. 

Second, if you attempt to optimize solely for accuracy of finite-time calculations _to experiment directly_, you will never get anywhere. Any choice you make (force field, calculation strategy, collision rate) will impact these finite-time calculations in uncontrollable and unknowable ways. We will never make any progress in improving either the force field, the protocol, or the calculation strategy because there is absolutely no way to separate any of these effects. Progress becomes completely impossible because all modifications are a random walk through calculation space.